core/src/main/scala/dagr/core/tasksystem/Pipe.scala

```diff
@@ -0,0 +1,182 @@
+  * before connecting them with pipes.
+  *
+  * See also the traits [[PipeIn]] and [[PipeOut]] which provide a simplified way to implement
+  * a piping task that only receives on or only emits to a pipe.
+  *
+
+  * @tparam In a symbolic type representing the kind of data the task can accept
+  * @tparam Out a symbolic type representing the kind of fdata the task emits
+  */
+trait Pipe[In,Out] extends ProcessTask {
+  /**
+    * Generates or extends a pipe chain by connecting this task to the next task via a pipeline.
+    *
+    * @param next a Pipe enabled task that should receive this pipe's output as its input
+    * @tparam In2 the type of input expected by the next task. Must be the same or a supertype
+    *             of this pipe's output.
+    * @tparam Out2 the type of output of the next task, which becomes the output type of the
+    *              returned pipe chain
+    * @return a new Pipe object that is the result of connecting the two pipes
+    */
+  def |[In2 >: Out,Out2](next: Pipe[In2,Out2]): Pipe[In,Out2] = {
+    Pipes.chain(this, next)
+  }
+
+  /** Returns the first task in the pipe. */
+  private[tasksystem] def left: Pipe[In,_] = this
+  /** Returns the last task in the pipe. */
+  private[tasksystem] def right: Pipe[_,Out] = this
+  /** Returns the ordered set of Pipes/tasks in the chain from first to last inclusive. */
+  private[tasksystem] def ordered: Seq[Pipe[_,_]] = List[Pipe[_,_]](this)
+}
+
+/** A simplified trait for sink tasks that can receive data via a pipe, but cannot pipe onwards. */
+trait PipeIn[In] extends Pipe[In,Nothing]
+
+/** A simplified trait for generate tasks that can pipe out, but cannot recieve data from a pipe. */
+trait PipeOut[Out] extends Pipe[Nothing,Out]
+
+
+///////////////////////////////////////////////////////////////////////////////
+// Section: Objects and private utility classes to make it all work
+///////////////////////////////////////////////////////////////////////////////
+object Pipes {
+  /**
+    * A class that represents an empty pipe, that allows for easier construction of pipes with conditional branches.
+    * Must remain private and only be intantiated by Piping.empty().
+    */
+  private class EmptyPipe[T] extends Pipe[T,T] {
+    override def args: Seq[Any] = throw new IllegalStateException("Should not reach here.")
+    override def pickResources(availableResources: ResourceSet): Option[ResourceSet] = throw new IllegalStateException("Should not reach here.")
+  }
+
+  /**
+    * An immutable class representing a chain of tasks that can be piped together. Currently only supports chains
+    * with up to a maximum of one non-FixedResources task, in order to keep the resource management code simple.
+    *
+    * The constructor does not check types as thoroughly as one might like, instead this is done in
+    * [[Pipes.chain()]], which is the only place the constructor is called from.
```

Comment:
I think it would be pretty easy to check in the constructor, given some modification to `Pipe`.  This works in my Repl:

```
import reflect.runtime.universe._
import scala.reflect.runtime.{universe => ru}
import scala.reflect._

val mirror: ru.Mirror = scala.reflect.runtime.currentMirror

class Pipe[Left: ClassTag, Right: ClassTag] {
  def left = classTag[Left]
  def right = classTag[Right]
}

def checkPipes(first: Pipe[_,_], second: Pipe[_,_]): Boolean = {
  first.right == second.left
}

val pipes = List[Pipe[_,_]](new Pipe[Int, String], new Pipe[String, Char], new Pipe[Char, Double], new Pipe[Int, Int])
for (i <- 1 until pipes.length) {
  println(s"Comparing ${i-1} and $i: " + checkPipes(pipes(i-1), pipes(i)).toString())
}
```